File: infosnippetplugin/trunk/infosnippet/infosnippet.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2021 Clemens
# All rights reserved.
#
# This software is licensed as described in the file COPYING, which
# you should have received as part of this distribution.

from trac.config import *
from trac.core import Component, implements
from trac.web.api import IRequestFilter
from trac.web.chrome import (
    ITemplateProvider, add_script, add_script_data, add_stylesheet)


class InfoSnippetComponent(Component):
    """
    This component uses Javascript to provide a text information box (snippet) with basic ticket infos like title and URL.
    """

    implements(IRequestFilter, ITemplateProvider)

    conf_section = ConfigSection('infosnippet',
      """
      This config section belongs to the Info-Snippet plugin.
      Its purpose is to gather basic information (like title and ULR) and copy it to the clipboard. 
      """)

    nav_option = ChoiceOption('infosnippet','nav',
      ['all','ticket','wiki','none'],
      """Specifies if the 'INFO' menu item shall appear in the context navigation menu at top of the page.
      Clicking this menu option will not navigate but instead copy the information snippet (as text) into the system clipboard.

      Valid options are: 
      - `ticket` only for ticket pages
      - `wiki` only for wiki pages
      - `all` for both of them
      - `none` neither of them
      """)

    box_option = ChoiceOption('infosnippet','box',
      ['all','ticket','wiki','none'],
      """Specifies if the 'Info Snippet' box shall appear near the bottom of the page.
      This is a box containing elemental information like title and URL for instance. 

      Valid options are: 
      - `ticket` only for ticket pages
      - `wiki` only for wiki pages
      - `all` for both of them
      - `none` neither of them
      """)

    # IRequestFilter methods

    def pre_process_request(self, req, handler):
        return handler

    def post_process_request(self, req, template, data, content_type):

        if template in ('ticket.html'):

            ticket = data['ticket']
            if ticket is not None:
                self.log.debug("Preparing InfoSnippet plugin for ticket.")
                add_script_data(req, info={
                          'navoption':self.nav_option,
                          'boxoption':self.box_option,
                          'projectname':self.env.project_name,
                          'ticketid':str(ticket.id),
                          'ticketsummary':ticket['summary']})
                add_script(req, 'info/infosnippet.js')
                add_stylesheet(req, 'info/infosnippet.css')

        if template in ('wiki_view.html'):
            self.log.debug("Preparing InfoSnippet plugin for wiki.")
            page = data.get('page')
            if page is not None:
                self.log.debug(page.name)
           
                add_script_data(req, info={
                          'navoption':self.nav_option,
                          'boxoption':self.box_option,
                          'projectname':self.env.project_name,
                          'page':page.name})
                add_script(req, 'info/infosnippet.js')
                add_stylesheet(req, 'info/infosnippet.css')

        return template, data, content_type

    # ITemplateProvider methods

    def get_htdocs_dirs(self):
        from pkg_resources import resource_filename
        return [('info', resource_filename(__name__, 'htdocs'))]

    def get_templates_dirs(self):
        return []

File: infosnippetplugin/trunk/infosnippet/htdocs/infosnippet.js
// TRAC TicketInfo-Plugin
//
// Copyright (C) 2021 Clemens
// All rights reserved.
//
// This software is licensed as described in the file COPYING, which
// you should have received as part of this distribution.

jQuery(function($) {

  // For the URL we are intentionally not using window.location.href 
  // because we want only the pure URL without local #anchors.
  var url = window.location.protocol + '//' + window.location.hostname + window.location.pathname;

  var tooltip = 'Copy info snippet text to clipboard';

  // We have a hidden Acknowledge text which appears only temporarily when the button is clicked.
  $ack = $('<span>')
    .addClass('infosnippet-ack')
    .attr('style','display: none; padding-right: 0.5em;')
    .text('copy to clipboard completed');

  
  CreateNav(url,$ack.clone(),tooltip);
  CreateBox(url,$ack.clone(),tooltip);
})


// create a "Info" context navigation menu item at top of page. It will copy to clipboard.
function CreateNav (url,ack,tooltip) {

  if ($('#infosnippet-nav').length == 0) { // only if not yet exists 
  if (info.navoption=='all' || info.navoption=='ticket' || info.navoption=='wiki') { 
  
    var $navbutton = $('<li/>');
    $("#ctxtnav ul li:first").after($navbutton);
    
    ack.appendTo($navbutton);
    $("<a>")
      .attr('href', "#infosnippet")
      .attr('id', 'infosnippet-nav')
      .attr("title", tooltip)
      .text("Info")
      .appendTo($navbutton);
      
    // register on-click handler 
    $navbutton.click(function() {
      $('#infosnippet').show(); // make sure the box is shown otherwise cannot copy to clipboard
      CopyToClipboard('infosnippet-text');
      if (info.boxoption!='all' && info.boxoption!='ticket') { 
        $('#infosnippet').hide(); // hide again if disable in config
      }
      $(".infosnippet-ack").show().delay(5000).fadeOut();
      return false;
    });
  }}
}

// Create a "Info Snippet" box at bottom of page
function CreateBox (url,ack,tooltip) {

  var $box;

  if ($('div.ticket').length){
  
    if ($('#infosnippet').length ) { return;} // skip if already exists
   
    // The "Ticket Info" box.
    $box = $('<fieldset>')
      .attr('id','infosnippet')
      .addClass('ticket');
      
    // Hide the box unless enabled in the config.
    // Note that we need the box (invisible or not), because want to copy its content to the clipboard. 
    $box.hide();
    if (info.boxoption=='all') { $box.show(); }
    if (info.boxoption=='ticket') { $box.show(); }
      
    $('<legend>').text('Ticket Info').appendTo($box);

    // the actual info snippet text
    $info= $('<p>')
      .attr('id','infosnippet-text')
      .append(info.projectname + ' Ticket #' + info.ticketid + '</br>'+ info.ticketsummary + '</br>')
      .appendTo($box);
    // URL of current page is part of the info snippet text.
    $("<a>")
      .attr('href',url)
      .append(url)
      .appendTo($info);
      
    // The fieldset is inserted right after the ticket properties
    $("div.ticket #properties").after($box);
  }
  else if ($('#content.wiki').length){
  
    if ($('#infosnippet').length ) { return;} // skip if already exists

    $box = $('<div>')
      .attr('id','infosnippet')
      .addClass('wiki');
      
    // Hide the box unless enabled in the config.
    // Note that we need the box (invisible or not), because want to copy its content to the clipboard. 
    $box.hide();
    if (info.boxoption=='all') { $box.show(); }
    if (info.boxoption=='wiki') { $box.show(); }
    
    // the actual info snippet text
    $info= $('<p>')
      .attr('id','infosnippet-text')
      .append(info.projectname + ' ' + info.page + '</br>')
      .appendTo($box);
    // search for the first head line in this page
    headline=$('#content.wiki .wikipage :header');
    if(headline.length){$info.append(headline[0].innerText+'</br>');}
    // URL of current page is part of the info snippet text.
    $("<a>")
      .attr('href',url)
      .append(url)
      .appendTo($info);
      
    // The infobox is inserted right after the ticket properties
    $("#content.wiki #attachments").after($box);
  }
  else {
    return;
  }
  
  $btn=$('<div>')
    .attr('id','infosnippet-button')
    .append(ack)
    .prependTo($box);
  $('<a>')
    .attr('href','')
    .text('Copy to Clipboard')
    .addClass('button')
    .attr("title", tooltip)
    .appendTo($btn);
  
  // register on-click handler 
  $btn.click(function() {
    CopyToClipboard('infosnippet-text');
    $(".infosnippet-ack").show().delay(5000).fadeOut();
    return false;
  });

}

// copy the text from a DIV container (given by ID) into the system clipboard
function CopyToClipboard (containerid) {
  if (document.getElementById(containerid) == null) return;
  var dummy= document.createElement('textarea')
  dummy.id = 'temp_element';
  dummy.style.height = 0;
  dummy.style.border = 0;
  document.body.appendChild(dummy);
  dummy.value = document.getElementById(containerid).innerText; // copy text from container into textarea
  document.getElementById("temp_element").select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
}

